### PR TITLE
added the edit mode functionality

### PR DESCRIPTION
**Pull request by jaspalsinghsaini**
* Added a button to enter/exit the edit mode
* For now, entering the edit mode stops scanning and enables directly clicking on buttons in the grid
* Add a unique tag for each button in the TouchGrid

### CODE REVIEW

**Comment by rgerd** [comment]
Excellent job Jaspal! Just a couple things I wanted to mention:
- I'm taking advantage of your PR just a little to inject some changes that reduce technical debt:
    - [I moved the array of action cards to the `VocabCard` class](https://github.com/rgerd/SwitchSpeak/blob/59397009ecb461623f9ecea6221e1cdda655cbdf/SwitchSpeak-iOS/SwitchSpeak-iOS/VocabCardStruct.swift#L62). This gives us a nice central place to modify the list of action buttons. [We use the array here](https://github.com/rgerd/SwitchSpeak/blob/0e4ab02d778cefc9c0c2079fbcca80fecf211958/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchSelection.swift#L45-L55) (i.e. as opposed to hard-coding '4' for the number of action buttons).
    - I replaced your passing down the `switchButton` [with a reference to the global singleton `TouchSelectionViewController` variable](https://github.com/rgerd/SwitchSpeak/blob/0e4ab02d778cefc9c0c2079fbcca80fecf211958/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchGrid.swift#L208). This is much simpler than having to keep track of a variable handed down through levels of abstraction.
    - [As you can see here](https://github.com/rgerd/SwitchSpeak/pull/53/commits/0e4ab02d778cefc9c0c2079fbcca80fecf211958#diff-c2e5305604f6be78cd2c03e5b32938ecL128), we were passing a `ButtonNode` to the breadcrumb manager, but immediately extracting the card data and never using the button node, so I decided to change that to just passing the button node's card data. This is especially convenient for a reason that will be made clear soon enough.
     - [I changed the name of `isEditMode` to `editing`](https://github.com/rgerd/SwitchSpeak/pull/53/commits/0e4ab02d778cefc9c0c2079fbcca80fecf211958#diff-b5ebf9aefe9d59cc3373cbd11a70ba54L21). It's just a little more linguistically natural. I also [changed `SwitchEditMode` to `toggleEditingMode`](https://github.com/rgerd/SwitchSpeak/pull/53/commits/0e4ab02d778cefc9c0c2079fbcca80fecf211958#diff-ed28cd923cec2e2141af7bbcac9ae446L29) for a similar reason, although I would also prefer if you stuck to camelCase for function names. :)
    - [I also replaced the hard-coded dummy node title with the dynamically defined one from `EmptyVocabCard`](https://github.com/rgerd/SwitchSpeak/pull/53/commits/0e4ab02d778cefc9c0c2079fbcca80fecf211958#diff-b5ebf9aefe9d59cc3373cbd11a70ba54L176). Gotta get rid of things like that when we can.
- I also modified your implementation of programmatically setting the `touchUpInside` event on our buttons. Here is yours ([part 1](https://github.com/rgerd/SwitchSpeak/blob/66f7b5cf588bc6e89cd96f657476241309512fc2/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchSelection.swift#L106-L145), [part 2](https://github.com/rgerd/SwitchSpeak/blob/66f7b5cf588bc6e89cd96f657476241309512fc2/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchGrid.swift#L220-L243)), and [here is mine](https://github.com/rgerd/SwitchSpeak/blob/0e4ab02d778cefc9c0c2079fbcca80fecf211958/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/ButtonNode.swift#L88-L119). You and Ning had a similar idea for this–find the nodes after building the tree and assign their actions. I think there is a simpler way. [What if we just assigned the actions as the button nodes are being built?](https://github.com/rgerd/SwitchSpeak/blob/0e4ab02d778cefc9c0c2079fbcca80fecf211958/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/ButtonNode.swift#L69) This way, we don't have to search for them, since they're right there in `self`! I moved the code for pushing cards to the breadcrumb stack [into a function called `selectCard`](https://github.com/rgerd/SwitchSpeak/blob/edit-mode/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchSelection.swift#L77-L86), and since tapping a card should be the same as selecting it with the switch button, [we can just call the button's `select` function](https://github.com/rgerd/SwitchSpeak/blob/edit-mode/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchSelection.swift#L74). While your method was totally valid, I think this might be easier to work with in the long-run, since we have less code and [it scales easily with different card types](https://github.com/rgerd/SwitchSpeak/blob/0e4ab02d778cefc9c0c2079fbcca80fecf211958/SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/ButtonNode.swift#L89). Please let me know what you think. Since this is such a big change I will hold off on merging until I get your approval.

**Comment by jaspalsinghsaini** [comment]
Those are some excellent design choices you have mentioned. 

* I had never used the singleton design pattern before, and I think that is a good choice for us given that we have only one instance of the `TouchSelectionViewController` class. Not yet sure if we will ever need two object of this class, hence I think we can stick to this design pattern for now.

* Please feel free to change any function/variable names given by me....You certainly have a better taste for selecting more intuitive identifier names. Was never happy with `switchEditMode`....I guess "toggle" was the word I was looking for :)

* I think I did try adding actions to the button nodes while building, but I found it to be challenging. If I understand it correctly, your use of the singleton design pattern makes this job a lot more easier and cleaner to implement. Excellent observation!!!

I also get to learn petty of software engineering terminologies from your PRs....got to learn "technical debt" on this occasion :)

Excellent job with this PR! Built and run the app too...and it works fine. 👍  for merging it into master.

File: SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchSelection.swift [hunk 1]
```diff
@@ -42,17 +42,17 @@ class TouchSelection {
         let (rows, cols) = settings.getGridSize()
         let gridSize:Int = rows * cols
         
-        let lastIndex:Int = min(cards.count, pageOffset + gridSize - 4) - 1
+        let lastIndex:Int = min(cards.count, pageOffset + gridSize - VocabCard.actionCards.count) - 1
```
Comment (jaspalsinghsaini) [inline SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchSelection.swift]: Always nice to remove these magic numbers 👍

**Comment by rgerd** [comment]
Thank you @jaspalsinghsaini! Merging away

**Pull request merged**